### ISSUE TITLE
System Design

### ISSUE THREAD

**Issue by timlinux** [kartoza/GEEST2#300]
# System Design Document: GEEST2

---

## 1. Introduction and Overview

This system performs geospatial analysis on small nations, focusing on women’s empowerment. Multiple workflows are run in parallel, with each indicator processed independently, and the results stored in a **GeoPackage**.

---

## 2. Architecture Overview

We follow a hierarchy of **dimensions**, **factors**, and **indicators**, with values aggregated at each level. The system efficiently processes these layers using **parallel tasks** on a single machine.

```mermaid
classDiagram
    class dimension {
      +id: string
      +name: string
    }
    class factor {
      +id: string
      +name: string
      +weight: float
    }
    class indicator {
      +id: string
      +name: string
      +weight: float
    }

    dimension "1" --> "1..*" factor
    factor "1" --> "1..*" indicator
```

---

## 3. Study Area and Grid System

The system generates a **100m x 100m grid** for the study area, using the boundaries of the relevant regions or islands. This grid is stored in the **study_area_grid** table, with each grid cell corresponding to a polygon.

- Grid cells are created over land based areas only
- Grid cells are assigned Likert-scale values (0-5) based on workflows.

```mermaid
graph TD;
    A[Study Area] --> B[Bounding Box]
    B --> C[100m x 100m Grid]
    C --> D[STUDY_AREA_GRID]
```

- **STUDY_AREA_GRID** stores each grid cell and its calculated values for each indicator, factor, and dimension.

---

### 4.1 Output Table Structure (study_area_grid)

The **study_area_grid** table is stored in a **GeoPackage** and includes:

- **Grid cells** (100m x 100m polygons).
- **Columns for each indicator**: Prefixed with **i_**.
- **Columns for each factor**: Prefixed with **f_**.
- **Columns for each dimension**: Prefixed with **d_**.

All values in these columns are stored as **integers** (from 0 to 5, following the Likert scale).

| id  | grid_cell | i_literacy | i_health | f_education | d_empowerment |
| --- | --- | --- | --- | --- | --- |
| 1   | POLYGON((...)) | 4   | 3   | 3   | 4   |
| 2   | POLYGON((...)) | 5   | 4   | 4   | 5   |
| ... | ... | ... | ... | ... |     |

### 4.2 Bounding Boxes Table (study_area_bboxes)

The **study_area_bboxes** table stores the **bounding boxes** for each individual polygon used during grid creation, aligned to the 100-meter grid. This table serves as a reference for the spatial extent of the study areas.

| id  | bbox_polygon | area_name | area_id |
| --- | --- | --- | --- |
| 1   | POLYGON((...)) | Island_1 | 101 |
| 2   | POLYGON((...)) | Island_2 | 102 |
| ... | ... | ... | ... |

- **bbox_polygon**: Polygon geometry representing the bounding box of the area being analyzed.
- **area_name**: The name of the specific area or island.
- **area_id**: A unique identifier for each study area.

### 4.3 Study Area Boxes Table (study_area_boxes)

The **study_area_boxes** table contains a similar structure to **study_area_bboxes**, but instead of representing overall bounding boxes for the entire region, this table stores the **bounding boxes** for each individual polygon that the system analyzes.

| id  | polygon_box | area_name | area_id |
| --- | --- | --- | --- |
| 1   | POLYGON((...)) | Island_1 | 101 |
| 2   | POLYGON((...)) | Island_2 | 102 |
| ... | ... | ... | ... |

---

## 8. Workflow Optimization and Parallelization

We accelerate the workflow processing by **parallelizing indicator computations**. Each indicator is processed by a separate task that runs concurrently, allowing the system to handle multiple indicators efficiently.

### 8.1 Parallel Task Execution for Indicators

In this design, **each indicator** has its own workflow task. These tasks run in parallel, updating the **study_area_grid** table independently. For example, the **literacy indicator** and the **health indicator** will be processed by separate tasks, which both update their respective columns in the table.

```mermaid
graph LR
    I1[Indicator 1 Task] --> updateGeo1[Update i_literacy Column]
    I2[Indicator 2 Task] --> updateGeo2[Update i_health Column]
    I3[Indicator 3 Task] --> updateGeo3[Update i_economic Column]
    updateGeo1 --> finalGeo[Final GeoPackage]
    updateGeo2 --> finalGeo
    updateGeo3 --> finalGeo
```

---

### 8.2 Example of Parallel Indicator Processing

```mermaid
sequenceDiagram
    participant User
    participant RasterTask
    participant TravelTask
    participant GeoPackage

    User->>RasterTask: Start Raster Indicator Workflow
    User->>TravelTask: Start Travel Time Indicator Workflow
    RasterTask->>GeoPackage: Update i_raster_indicator Column
    TravelTask->>GeoPackage: Update i_travel_indicator Column
    GeoPackage->>User: Both Workflows Complete
```

---

### 8.3 Optimizing Workflow Execution

We use several optimization techniques to ensure efficient parallel execution:

```mermaid
graph TD
    task1[Indicator Workflow 1] --> spatial[Spatial Indexing]
    task2[Indicator Workflow 2] --> cache[In-memory Caching]
    task3[Indicator Workflow 3] --> batch[Batch Updates to GeoPackage]
    spatial --> finalGeo[Efficient GeoPackage Updates]
    cache --> finalGeo
    batch --> finalGeo
```

---

### 8.4 Error Handling and Workflow Recovery

```mermaid
graph LR
    start[Start Indicator Task] --> checkpoint[Checkpoint Progress]
    checkpoint --> logErrors[Log Errors]
    logErrors --> retry[Retry Task if Needed]
    retry --> success[Task Completed Successfully]
```

---

### Final System Workflow Integration

Here’s how the overall system executes indicator workflows in parallel, then aggregates the results into **factors** and **dimensions**:

```mermaid
graph LR
    user[User Input] --> indicator[Parallel Indicator Workflow Tasks]
    indicator --> factor[Compute Factor Values from Indicators]
    factor --> dimension[Compute Dimension Values from Factors]
    dimension --> updateGeo[Update GeoPackage]
    updateGeo --> taskComplete[All Tasks Completed]
```

---

## Conclusion

By parallelizing indicator workflows and running each one independently, the system efficiently processes multiple indicators at the same time on a single machine. This method avoids grid-based partitioning and allows each indicator to be processed as its own unit, updating the corresponding columns in the **GeoPackage**. Combined with optimizations like spatial indexing, caching, and batch updates, the system ensures fast and efficient geospatial analysis.
